core/src/unprocessed_packet_batches.rs

```diff
@@ -91,6 +102,42 @@ impl UnprocessedPacketBatches {
         UnprocessedPacketBatches(VecDeque::with_capacity(capacity))
     }
 
+    /// Insert new `deserizlized_packet_batch` into inner `VecDeque<DeserializedPacketBatch>`,
+    /// If buffer is at Max limit, packets (not packet_bacth) will be dropped, starting from
+    /// the lowest priority until an empty batch is located, then swap it with new batch;
+    /// Otherwise, new_batch will be pushed into the end of VecDeque;
+    /// returns tuple of (number of batch dropped, number of packets dropped)
+    pub fn insert_or_swap_batch(
+        &mut self,
+        deserialized_packet_batch: DeserializedPacketBatch,
+        batch_limit: usize,
+    ) -> (Option<usize>, Option<usize>) {
+        if deserialized_packet_batch.unprocessed_packets.is_empty() {
+            return (None, None);
+        }
+
+        if self.len() >= batch_limit {
+            let (_, dropped_batches_count, dropped_packets_count) =
+                self.swap_packet_by_priority(deserialized_packet_batch);
```

Comment:
When there are a lot of incoming packets during spikes, I see this pathway being hit repeatedly in the metrics, meaning it can't be expensive to run.

This currently a pretty expensive operation of figuring out the fees for *every* transaction in the queue, sorting them, dropping the unnecessary ones.

Maybe it would be better to change the queue to support ordering by the fee-per-cu on insertion, maybe a `BTreeMap`?

Reply:
This is a good point, I paid attention on bench test and perf test. Tests didn't show noticeable degrading, but the concern is valid. 

If the buffered entry is `packet_batch` (which I'd like to avoid changing it to minimize the change scope), I'm not sure how BTreeMap could help here. 

Another constrain is fee-per-cu is `bank` dependent, it is not simply derived from transaction itself (at least at the moment), rather it depends on current bank's fee_structure and lamports_per_signature. 

Amount the added overhead in order to be able to prioritize by fee, I think compute fee per transaction perhaps the most expensive. Sorting by fee (utilized a BTreeMap) might be second; dropping should cost very little (drop a `packet` is just remove its index from `unprocessed vec`, drop a batch is a HashMap::Swap op).

Mentioned somewhere early, I intended to increase banking_stage non-vote threads from 2 to 4 to compensate added overhead, which also automatically reduce each thread's buffer size (`= TOTAL_BUFFERED_PACKETS/number_of_thredas`), that'd help on reduce sorting cost a bit. (I tested it before, the benching show significant improvement with 6 threads, more test should be done ofc)

Having said that, I think to cache computed fee-per-cu for transactions for up to X slots would help. Ideally X is `1`, so it computes fee-per-cu for each transaction once per bank, still would save a lot during spikes. Or we can increase X if we don't care to be super precise, or performance is prove be too bad during spikes. I'll start work on this.

Reply:
[commit](d0a0bf669e505e8e978b18a12f3182a070c13971) adds cache for fee-per-cu for up to 1 slot. Going to add some metrics

Reply:
> Tests didn't show noticeable degrading, but the concern is valid.

Hmm I think to test this drop path properly, we would have to spam packets such that each thread hits the buffer limit, i.e. somewhere close to 500k tps, which I don't think any of the tests currently do

> Another constrain is fee-per-cu is bank dependent, it is not simply derived from transaction itself (at least at the moment), rather it depends on current bank's fee_structure and lamports_per_signature.

I think if we follow your suggestion here: https://github.com/solana-labs/solana/pull/23841#discussion_r833769663, it won't matter too much, we just need to pick one consistent bank so that everyone is relatively equal to one another.

> If the buffered entry is packet_batch (which I'd like to avoid changing it to minimize the change scope), I'm not sure how BTreeMap could help here.

Yeah this is a pain, we really want to be able to weigh each packet separately in a `min-max-heap`. This will help too when we need to find the highest priority transaction to process, which my 🔮  is telling me is going to be the follow-up to this 😉 

With our current grouping of multiple packets into `PacketBatch` there's also a correctness problem when a high priority transaction is grouped into the same batch as a low priority batch. This is exemplified by compromises like this: https://github.com/solana-labs/solana/blob/d0a0bf669e505e8e978b18a12f3182a070c13971/core/src/unprocessed_packet_batches.rs#L400 where a potentially high fee transaction doesn't make it into the buffer because a bunch of low transaction transactions were grouped into the same batch as high priority transactions in a `PacketBatch`


I think we should consider ripping off the bandaid and switch to operating on individual `Packet` objects here.

Reply:
> I think we should consider ripping off the bandaid and switch to operating on individual `Packet` objects here.

Yeah, the benefits of dealing with individual packets is obvious, particularly from the angle we are in right now 😃 ; Just concerned if there are downsides are not apparent. Better to do this separately, imo.

Reply:
> there's also a correctness problem when a high priority transaction is grouped into the same batch as a low priority batch. This is exemplified by compromises like this:

Using prioritized `PacketLocator`s, such batching shouldn't an issue anymore - high priority packet is up front in prioritized `Vec<PacketLocator>` while low priority ones from same batch will be at the bottom of queue. 

Hmm, the `warn!` scenario shouldn't really happen. In that code, all unprocessed packets are indexed by priority from low to high, then packet by packet they are removed from `unprocessed` list (eg., dropped) until a batch without unprocessed packet (eg., empty batch) is found. I can't think of a case when all unprocessed packets are dropped yet can't still find an empty batch. I added `warn!` there for testing, but I don't expect it's being hit tbh.

Reply:
> Better to do this separately

Yeah agreed should be in a PR that lands before this one

>Hmm, the warn! scenario shouldn't really happen. In that code, all unprocessed packets are indexed by priority from low to high, then packet by packet they are removed from unprocessed list (eg., dropped) until a batch without unprocessed packet (eg., empty batch) is found.

I see, but it still has the potential to sacrifice some highly prioritized packets. For instance if I had two packet batches:

```
1. PacketBatch{1, 1, 1, 1, 100}
2. PacketBatch{99,99,99,99}
```

Assume an extreme case where the max capcity is 1. Then the second batch will be dropped, even though it contained many of the highest priority transactions because of unlucky batching.

Reply:
Yep, in that extreme scenario, the `99`s would be chopped. I am sure there are benefits with "batch", but its granularity is sacrificed somehow.

Reply:
One potential game plan could be: 
1. landing a version of QoS that's built on top of packet_batches, 
2. iterate back to replace packet_batch with pakctes,
3. implement Carl/Trent proposed singleton scheduler,
4. implement new version of QoS based on all above.

Reply:
I think this current code for dropping batches once the limit is hit is equivalent and could be simplified to:

1. Weighting each `PacketBatch` by its heaviest packet. Examples: `PacketBatch{1, 1, 1, 1, 100}` would have weight 100 and `PacketBatch{99,99,99,99}` would have weight `99`. 
2. Dropping the "smallest" such `PacketBatch` 

Then we can use a min-max heap: https://docs.rs/min-max-heap/latest/min_max_heap/ to find the smallest to drop

Reply:
Hmm, but then all these `99` packets will be dropped while `1`s are let in. I think this is the scenario to be addressed, at least for the short term. 

I agree that using sorted container helps, but each batch's weight changes whenever a packet in it is processed. So it still needs to scan packet by packet continuously to update each batch's weight, not sure if there is a _large_ saving from what's in the PR, which is to flat out packets from buffered batches, then dropping packets from lowest priority. 

Perhaps it is the best to come up a performance test, so we have some concrete to look at it. I'll create an issue for it.

Reply:
issue #24006 ^^

Reply:
@taozhu-chicago yeah you're right, thats a subtlety I missed. One other thing we could do is create a 

```
struct PacketWeight {
    original_packet_batch: Rc<PacketBatch>,
    packet_weight: u64
}
```

and create a min-max heap of these, removing the `original_packet_batch` when all of its `PacketWeight` objects have been removed.

Reply:
Mmm interesting, I evaluated options of creating and updating some sort of "master index" using btree_map, vs lazily creating disposable index when needed. Former requires op for each received packets, later only acts when need to drop batches. Looked at the metrics, it seems to receives thousand of packets while only drop handful of batches. So I chose the lazy approach to implement. But that most likely during normal condition.

The min-max heap can be useful in other parts as well. I think what is the most efficient way of indexing packets much depends on the traffic condition. I will copy your idea to issue #24006

Reply:
Summary of some benchmarks I made here: https://github.com/solana-labs/solana/issues/24006#issuecomment-1085434208

Reply:
> This is a good point, I paid attention on bench test and perf test. Tests didn't show noticeable degrading, but the concern is valid.
> 
> If the buffered entry is `packet_batch` (which I'd like to avoid changing it to minimize the change scope), I'm not sure how BTreeMap could help here.
> 
> Another constrain is fee-per-cu is `bank` dependent, it is not simply derived from transaction itself (at least at the moment), rather it depends on current bank's fee_structure and lamports_per_signature.
> 
> Amount the added overhead in order to be able to prioritize by fee, I think compute fee per transaction perhaps the most expensive. Sorting by fee (utilized a BTreeMap) might be second; dropping should cost very little (drop a `packet` is just remove its index from `unprocessed vec`, drop a batch is a HashMap::Swap op).
> 
> Mentioned somewhere early, I intended to increase banking_stage non-vote threads from 2 to 4 to compensate added overhead, which also automatically reduce each thread's buffer size (`= TOTAL_BUFFERED_PACKETS/number_of_thredas`), that'd help on reduce sorting cost a bit. (I tested it before, the benching show significant improvement with 6 threads, more test should be done ofc)
> 
> Having said that, I think to cache computed fee-per-cu for transactions for up to X slots would help. Ideally X is `1`, so it computes fee-per-cu for each transaction once per bank, still would save a lot during spikes. Or we can increase X if we don't care to be super precise, or performance is prove be too bad during spikes. I'll start work on this.

Hopefully this isn't too off-topic here, but given that y'all are trying to save as much time as possible, would the sorting by stake-weight be necessary and achieving what is desired? My understanding from when Anatoly spoke ~2 months ago is that he wanted even low-staked validators to have a guaranteed spot in blocks and there to be "fair" MEV opportunities I think (though validators will probably end up doing their own client prioritization in the future for these opportunities?). The problem I see is that if there is any shuffling/sorting of the queue by stake weight, and the queue is longer than what can fit in a block, it actually reduces the chances that smaller validators get transactions included in blocks (reduces the chances below their given stake weight % and increases inclusion of larger validators beyond their stake weight %). 

Scenario: 10,000 transactions are in queue but only 1000 will fit in block. 1000 transactions originate from top 33% of validators by stake weight. This means that only those transactions from top 1000 will make it in? I think the issue here would be that it gives large validators even more MEV power than they already have. If this isn't the desired behavior maybe time would be better spent on something else.

Reply:
Ahh right right, I got the part about the fee-per-cu. I was just thinking in the case where fee-prioritization is a constant --- it would put the smaller validators at that inherent disadvantage when queue exceeds block size. The only way for them to get any transaction inclusion would then be to submit transactions with higher fee-priority than the larger validators. I guess I just wasn't quite sure what the goal was of the stake weight prioritization if Quic happens to work. I think for both Quic and stake-weight priority it'll be something that validators end up tossing aside at some point anyway if MEV or block rewards makes them more $ by not following those rules.

Reply:
Yea, I can see your point. The original idea behind stake-weight is "validator has x% stake should have x% say on what transactions to be included in block". At what stage it to be implemented is more flexible. If Quic works, we can easily toss this out from banking_stage, i think.

Reply:
Hmm maybe the other thing that could be possible is that you give precedence to a certain sized set of immediately prior leaders, this would still be stake-weighted (because higher staked validators appear more often in schedule), but would allow for lower staked validators to appear in the ordering (since they'll pop up in the set too). Also, in the case that the prior leaders' transactions may have not all gone through, they still have some lasting prioritization beyond their leader slots (this is more loose than direct forwarding, but maybe helps?? dunno, lol).

Reply:
> Good question, but just want to point it out stake weight here plays a small part in prioritization. Transactions in buffer are prioritized by their fee-per-cu. Sender stakes are weighted to transactions have same fee-per-cu. Arguably, it is optional to using sender stake in leader prioritization, just an extra.

Coming back to this, given fee priority rate will be quite fine-grained as of now, I think any type of stake weight ordering is a bit pointless. I think you'll see a wide variety of priority except for maybe at the base level, but if I'm a bot I'll never send at that level, I'll add add a few micro lamports and send transactions across a range of priority which doesn't perceptibly cost me anything, then some other bot will add a few micro lamports in response until you never see priority fees at the exact same level. There won't be any beneficial stake weighted ordering that can occur at that point anymore. I think ordering by stake weight priority might just waste valuable time.

Reply:
> > Good question, but just want to point it out stake weight here plays a small part in prioritization. Transactions in buffer are prioritized by their fee-per-cu. Sender stakes are weighted to transactions have same fee-per-cu. Arguably, it is optional to using sender stake in leader prioritization, just an extra.
> 
> Coming back to this, given fee priority rate will be quite fine-grained as of now, I think any type of stake weight ordering is a bit pointless. I think you'll see a wide variety of priority except for maybe at the base level, but if I'm a bot I'll never send at that level, I'll add a few micro lamports and send transactions across a range of priority which doesn't perceptibly cost me anything, then some other bot will add a few micro lamports in response until you never see priority fees at the exact same level and things will be too dynamic to ever fall at the same exact priority rate. There won't be any beneficial stake weighted ordering that can occur at that point anymore. I think ordering by stake weight priority might just waste valuable time.

agree. Perhaps worth to see how it plays out.